storage/rocksdb/ha_rocksdb.cc

```diff
@@ -643,6 +659,60 @@ static int rocksdb_force_flush_memtable_and_lzero_now(
   return num_errors == 0 ? HA_EXIT_SUCCESS : HA_EXIT_FAILURE;
 }
 
+static void rocksdb_compact_lzero_now_stub(
+    THD *const thd MY_ATTRIBUTE((__unused__)),
+    struct SYS_VAR *const var MY_ATTRIBUTE((__unused__)),
+    void *const var_ptr MY_ATTRIBUTE((__unused__)),
+    const void *const save MY_ATTRIBUTE((__unused__))) {}
+
+static int rocksdb_compact_lzero_now(
+    THD *const thd MY_ATTRIBUTE((__unused__)),
+    struct SYS_VAR *const var MY_ATTRIBUTE((__unused__)),
+    void *const var_ptr MY_ATTRIBUTE((__unused__)),
+    struct st_mysql_value *const value) {
+  bool parsed_value = false;
+
+  if (mysql_value_to_bool(value, &parsed_value) != 0) {
+    return 1;
+  } else if (!parsed_value) {
+    // Setting to OFF is a no-op and this supports mtr tests
+    return HA_EXIT_SUCCESS;
+  }
+
+  return rocksdb_compact_lzero();
+}
+
+static void rocksdb_force_flush_memtable_and_lzero_now_stub(
+    THD *const thd MY_ATTRIBUTE((__unused__)),
+    struct SYS_VAR *const var MY_ATTRIBUTE((__unused__)),
+    void *const var_ptr MY_ATTRIBUTE((__unused__)),
+    const void *const save MY_ATTRIBUTE((__unused__))) {}
+
+static int rocksdb_force_flush_memtable_and_lzero_now(
+    THD *const thd MY_ATTRIBUTE((__unused__)),
+    struct SYS_VAR *const var MY_ATTRIBUTE((__unused__)),
+    void *const var_ptr MY_ATTRIBUTE((__unused__)),
+    struct st_mysql_value *const value) {
+  bool parsed_value = false;
+
+  if (mysql_value_to_bool(value, &parsed_value) != 0) {
+    return 1;
+  } else if (!parsed_value) {
+    // Setting to OFF is a no-op and this supports mtr tests
+    return HA_EXIT_SUCCESS;
+  }
+
+  // NO_LINT_DEBUG
+  LogPluginErrMsg(INFORMATION_LEVEL, ER_LOG_PRINTF_MSG,
+                  "RocksDB: Manual memtable and L0 flush.");
+  rocksdb_flush_all_memtables();
+
+  // Try to avoid https://github.com/facebook/mysql-5.6/issues/1200
```

Comment:
If two threads both try to set rocksdb_compact_lzero=1 or rocksdb_force_flush_memtable_and_lzero_now=1, will they hit similar race?

Reply:
Note that by race I mean a performance issue, not a correctness issue:
1) request memtable flush
2) immediately request L0 -> L1 compaction
3) wait for both to finish

If L0 -> L1 starts before memtable flush finishes then there might still be SSTs in the L0.

The "hack", sleep for 1 second in rocksdb_force_flush_memtable_and_lzero now between memtable flush and L0->L1 compaction, reduces the window for the race, but doesn't prevent it. There are other ways for the race to occur and I am reluctant to try and prevent them as any solution requires usage of a robust solution for waiting for compaction/flush to finish. Such a robust solution doesn't exist today. Hopefully it will in a few weeks or months. But even then I am still reluctant to try too hard to make this perfect.

It is more than good enough if I fix the bug WRT to getting the base level for L0->L1 compaction